core/src/main/scala/kafka/controller/ControllerEventManager.scala

```diff
@@ -77,7 +77,7 @@ class ControllerEventManager(controllerId: Int,
   private val putLock = new ReentrantLock()
   private val queue = new LinkedBlockingQueue[QueuedEvent]
   // Visible for test
-  private[controller] val thread = new ControllerEventThread(ControllerEventThreadName)
+  private[controller] var thread = new ControllerEventThread(ControllerEventThreadName)
```

Comment:
This comment applies to `clearAndPut`:

```scala
  def clearAndPut(event: ControllerEvent): QueuedEvent = inLock(putLock) {
    queue.forEach(_.preempt(processor))
    queue.clear()
    put(event)
  }
```
I think there is a bug here where at most one event will be process twice. Once by `_preempt(processor)` and once by `doWork`. I think we can fix this concurrency bug if we use `LinkedBlockingQueue::drainTo`. E.g.

```scala
  def clearAndPut(event: ControllerEvent): QueuedEvent = {
    val events = ...;

    inLock(putLock) {
      queue.drainTo(events)
      put(event)
    }

    event.forEach(_.preempt(processor))
  }
```

Reply:
@jsancio updated the PR. i wrapped the entire method in a lock because the return type `QueuedEvent` can be retrieved by `put(event)`. the suggested code would require another variable to be initialized, then set inside the lock, then returned at the end. I thought it would be best to prefer simplicity and readability over performance as this only occurs in the start and end of `ControllerEventManager`'s lifecycle.